spec/controllers/mfa_confirmation_controller_spec.rb

```diff
@@ -1,46 +1,100 @@
 require 'rails_helper'
 
 describe MfaConfirmationController do
-  describe 'GET /reauthn' do
+  describe '#new' do
     it 'presents the password confirmation form' do
       stub_sign_in
 
       get :new
 
       expect(response.status).to eq 200
+      expect(session[:password_attempts]).to eq 0
+    end
+
+    it 'does not reset password attempts if already set' do
+      stub_sign_in
+      session[:password_attempts] = 1
+
+      get :new
+
+      expect(session[:password_attempts]).to eq 1
     end
   end
 
-  describe 'PUT /reauthn' do
+  describe '#create' do
     let(:user) { build(:user, password: 'password') }
 
     before do
       stub_sign_in(user)
+      session[:password_attempts] = 1
     end
 
     context 'password is empty' do
-      it 'redirects with error message' do
+      it 'redirects with error message and increments password attempts' do
         post :create, user: { password: '' }
 
         expect(response).to redirect_to(user_password_confirm_path)
         expect(flash[:error]).to eq t('errors.confirm_password_incorrect')
+        expect(session[:password_attempts]).to eq 2
       end
     end
 
     context 'password is wrong' do
-      it 'redirects with error message' do
+      it 'redirects with error message and increments password attempts' do
         post :create, user: { password: 'wrong' }
 
         expect(response).to redirect_to(user_password_confirm_path)
         expect(flash[:error]).to eq t('errors.confirm_password_incorrect')
+        expect(session[:password_attempts]).to eq 2
       end
     end
 
     context 'password is correct' do
-      it 'redirects to 2FA' do
+      it 'redirects to 2FA and resets password attempts' do
         post :create, user: { password: 'password' }
 
         expect(response).to redirect_to(user_two_factor_authentication_path(reauthn: true))
+        expect(session[:password_attempts]).to eq 0
+      end
+    end
+  end
+
+  describe 'password attempts counter' do
+    context 'max password attempts reached' do
+      it 'signs the user out' do
+        user = build(:user, :signed_up)
```

Comment:
would it make any sense to pull some of this repeated code into a `before` block?